cas-server-support-spnego/src/main/java/org/jasig/cas/support/spnego/web/flow/client/BaseSpnegoKnownClientSystemsFilterAction.java

```diff
@@ -0,0 +1,313 @@
+                        this.alternativeRemoteHostAttribute, userAddress);
+            }
+        }
+        return userAddress;
+    }
+
+    /**
+     * Alternative header to be used for retrieving the remote system IP address.
+     * @param alternativeRemoteHostAttribute the alternative remote host attribute
+     */
+    public final void setAlternativeRemoteHostAttribute(@NotNull final String alternativeRemoteHostAttribute) {
+        this.alternativeRemoteHostAttribute = alternativeRemoteHostAttribute;
+    }
+
+    /**
+     * Regular expression string to define IPs which should be considered.
+     * @param ipsToCheckPattern the ips to check as a regex pattern
+     */
+    public final void setIpsToCheckPattern(@NotNull final String ipsToCheckPattern) {
+        this.ipsToCheckPattern = Pattern.compile(ipsToCheckPattern);
+    }
+
+    protected final String getRemoteIp() {
+        return this.remoteIp;
+    }
+
+    @Override
+    public final String toString() {
+        return new ToStringBuilder(this)
+                .append("ipsToCheckPattern", this.ipsToCheckPattern)
+                .append("alternativeRemoteHostAttribute", this.alternativeRemoteHostAttribute)
+                .append("remoteIp", this.remoteIp)
+                .append("timeout", this.timeout)
+                .toString();
+    }
+
+    /**
+     * Set timeout (ms) for DNS requests; valuable for heterogeneous environments employing
+     * fall-through authentication mechanisms.
+     * @param timeout # of milliseconds to wait for a DNS request to return
+     */
+    public final void setTimeout(final long timeout) {
+        this.timeout = timeout;
+    }
+
+    /**
+     * Convenience method to perform a reverse DNS lookup. Threads the request
+     * through a custom Runnable class in order to prevent inordinately long
+     * user waits while performing reverse lookup.
+     * @return the remote host name
+     */
+    protected String getRemoteHostName() {
+        final ReverseDNS revDNS = new ReverseDNS(getRemoteIp());
+
+        final Thread t = new Thread(revDNS);
+        t.start();
+
+        try {
+            t.join(this.timeout);
+        } catch (final InterruptedException e) {
+            logger.debug("Threaded lookup failed.  Defaulting to IP {}.", getRemoteIp(), e);
+        }
+
+        final String remoteHostName = revDNS.get();
+        logger.debug("Found remote host name {}.", remoteHostName);
+
+        return StringUtils.isNotEmpty(remoteHostName) ? remoteHostName : getRemoteIp();
+    }
+
+    /**
+     *  Utility class to perform DNS work in a threaded, timeout-able way
+     *  Adapted from: http://thushw.blogspot.com/2009/11/resolving-domain-names-quickly-with.html.
+     *
+     *  @author Sean Baker sean.baker@usuhs.edu
+     *  @author Misagh Moayyed
+     *  @since 4.1
+     */
+    private static final class ReverseDNS implements Runnable {
```

Comment:
Good point. I think I'll move this down to the hostname based action class. It has no use to be exposed as a public API.